### PR TITLE
Determine data dimensions lazily on fit instead on init

### PR DESCRIPTION
**Pull request by timokau**
## Description

Continuation of #94.

This is a work in progress. I'm putting this out there since it is likely I won't have any time left to finish it this week.

Things to do:
- ~Unify all the changes. Initially I started of by just inferring the data dimensionality from the current data without storing it in an instance attribute. This requires more invasive refactorings however, since it means the dimension will only be available from the `fit` function and the `fit` function has to pass it along as parameters. It doesn't always work either, since sometimes the dimensionality is needed at prediction time and then it sometimes matters if the dimensionality of the data we're predicting on is different from the dimensionality of the data we have trained on. Therefore I started storing the dimensionality the current model was fit on in instance attributes. I should adapt the earlier changes to this style to increase uniformity and reduce the chance of error.~
- ~Verify that pre-commit hooks and the full test suite pass after every commit.~
- ~Find all initializations in which the dimensionality is passed to the estimator. Currently this doesn't throw an error since the estimators accept `**kwargs`. Still, the calls should be adapted.~
- ~Double-check if any more documentation needs to be modified to remove references to `n_objects` and `n_object_features`.~

## Motivation and Context

 The `scikit-learn` estimator API only allows parameters with default values for `__init__`. There are no sensible defaults for the data dimension. Further, its just unnecessary to fix the data dimension on init. The developer experience is much nicer when it can simply be inferred from the data passed to `fit`.

## How Has This Been Tested?

Ran the full test suite.

## Does this close/impact existing issues?

Impacts #94, #116 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: csrank/discretechoice/generalized_nested_logit.py [hunk 1]
```diff
@@ -92,12 +92,7 @@ def __init__(
         """
         self.logger = logging.getLogger(GeneralizedNestedLogitModel.__name__)
 
-        self.n_object_features = n_object_features
-        self.n_objects = n_objects
-        if n_nests is None:
-            self.n_nests = n_objects + int(n_objects / 2)
```
Comment (timokau) [inline csrank/discretechoice/generalized_nested_logit.py:98]: This looks like a bug to me. According to the documentation and the implementation in `nested_logit_model.py` we should drop the `n_objects` summand.

@prithagupta do you agree? (you added this code originally)

**Comment by timokau** [comment]
Okay, I think/hope I found and fixed all instances now and removed mentions of the data dimensionality from the init docstrings. Tests pass. Still needs cleanup.

File: csrank/objectranking/list_net.py [hunk 2]
```diff
@@ -90,9 +84,6 @@ def __init__(
                 [1] Z. Cao, T. Qin, T. Liu, M. Tsai and H. Li. "Learning to Rank: From Pairwise Approach to Listwise Approach." ICML, 2007.
         """
         self.logger = logging.getLogger(ListNet.__name__)
-        self.n_object_features = n_object_features
-        self.n_objects = n_top
-        self.n_top = n_top
```
Comment (timokau) [inline csrank/objectranking/list_net.py:95]: This may need review. I have assumed `n_top` is essentially equivalent to what is called `n_objects` in other estimators and can be inferred from the data as well. Is that correct? Why does it have a different name?

**Comment by timokau** [comment]
I have cleaned everything up and reviewed my changes. This is ready for another pair of eyes now.

This is quite a big change. The basic idea is to get rid of the required `n_obejcts` and `n_object_features` arguments in our Learner constructors. We can infer that information from the data that is passed to `fit`. As a consequence, some other initializations that depend on the dimensionality have to be delayed to `fit` as well. Technically its a breaking change, though in practice it doesn't make much of a difference since all our learners quietly "swallow" additional unused arguments to init anyways. To keep potential for breakage as small as possible, I kept changes minimally invasive. Its tempting to do some fixups along the way, but that would be harder to do right with so many changes in so many places.

The repetitiveness of some of the changes shows that there is a lot of duplicate code. I haven't introduced any new abstractions for now because of the risk of introducing errors and because some big refactor / partial rewrite is on the horizon anyway (either the tf2 update or a switch to pytorch).

I hope we can get this merged relatively quickly, since it touches a lot of code and will otherwise likely cause merge conflicts. Keep in mind that I've already reviewed all the changes myself, so a second review doesn't have to be too detailed. I have left two comments on things that may need more review in the diff.

We can ignore CodeCov. Looks like there is still some configuration issue. If anything, this should increase coverage by decreasing the lines of code.

File: csrank/discretechoice/generalized_nested_logit.py (hunk 1 above)
Comment (kiudee) [reply csrank/discretechoice/generalized_nested_logit.py:98]: Indeed, that seems to be a bug. We should probably open an issue about this and fix it on a separate pull request.

File: csrank/objectranking/list_net.py (hunk 2 above)
Comment (kiudee) [reply csrank/objectranking/list_net.py:95]: ListNet as defined in the original paper only calculates the top-1 ranking probability in order to define a suitable and efficient loss function. This later was generalized to top-k rankings.
The `n_top` here is the k and specifies how many objects to consider for the loss function.
This is why setting `n_top = n_objects` is not the same thing.

Comment (timokau) [reply csrank/objectranking/list_net.py:95]: Ah, okay. Its a bit confusing that the original code set `n_objects` to `n_top`, but then `n_objects` wasn't used anywhere either.

Since we still need to avoid mandatory init arguments, how should we handle this case? Would it be reasonable to default to `None` and then fill that in with `n_objects` on `fit`? Or, going with the initial formulation in the paper, default to `1` instead?

Comment (kiudee) [reply csrank/objectranking/list_net.py:95]: I would be in favor of a default value. `1` sounds fine.

Comment (timokau) [reply csrank/objectranking/list_net.py:95]: Okay, thanks for the feedback. I have implemented the default value of `1`, please have another look.

File: csrank/discretechoice/generalized_nested_logit.py (hunk 1 above)
Comment (prithagupta) [reply csrank/discretechoice/generalized_nested_logit.py:98]: In nested logit, we are taking default number of nests to be
self.n_objects = n_objects
        if n_nests is None:
            self.n_nests = int(n_objects / 2)
        else:
            self.n_nests = n_nests
 , while in generalized_nested_logit,
its
  self.n_nests = n_objects + int(n_objects / 2)

The nested logit model divides the objects into a mutually exclusive set so the nests should be less than number of objects if n_nests=n_objects, then it becomes a logit model.
While GenNestedModel, says every object to some extend is a part of a nest, there is some literature which suggests still the number of nests should be less than number of objects in the choice set. So, in principle it is a good idea to get the number of objects in a choice set.
